Bolt agent added to hello metadata

ConorNeo opened the pull request:
<!--
    Thanks for making the effort to create a Pull Request!
    Please read the comment in its entirety fist.
    If you haven't already, please sign our Contributor License Agreement at
    https://neo4j.com/developer/cla/
    Commits from accounts that have not signed the CLA will fail the CI and
    will not be reviewed. If you are a first-time contributor and have just
    signed the CLA, please include a note about this in the PR comment as some
    manual steps from our side are required.
-->

<!--
    Title: Bolt Agent
-->

<!--
    Description:
    Bolt Agent added to hello message when protocol is 5.3. When user agent is not set in previous versions it should be set 
    the bolt agent! 
-->

bigmontz committed:
Enables Deno specific tests in Core

Core features which are environment dependent will need to be implemented.
For testing this features, environment dependent tests need to be setup.

The `deno` tests can be run by `npm run test::deno` and they need to be located under `./packages/core/test/deno`.
The testkit pipeline is configured to run these tests only when configured for test `deno`.

ConorNeo committed:
Bolt agent added to hello metadata

bigmontz commented on packages/bolt-connection/src/bolt/bolt-protocol-v3.js:
```diff
@@ -78,7 +78,7 @@ export default class BoltProtocol extends BoltProtocolV2 {
     assertNotificationFilterIsEmpty(notificationFilter, this._onProtocolError, observer)
 
-    this.write(RequestMessage.hello(userAgent, authToken), observer, true)
+    this.write(RequestMessage.hello(userAgent === '' || userAgent == null ? boltAgent : userAgent, authToken), observer, true)
```
Should we treat `''` as null?

ConorNeo replied:
if i dont it causes a testkit test to fail

bigmontz commented on packages/bolt-connection/src/bolt/request-message.js:
```diff
@@ -191,6 +191,38 @@ export default class RequestMessage {
+   * @return {RequestMessage} new HELLO message.
+   */
+  static hello5x3 (userAgent, boltAgent, notificationFilter = null, routing = null) {
+    const metadata = { user_agent: userAgent, bolt_agent: boltAgent }
```
maybe we should omit `user_agent` if it's `null` or `undefined`.

ConorNeo replied:
yeah makes sense!

ConorNeo committed:
Version 1 fix for user agent being empty

bigmontz commented on packages/core/src/internal/bolt-agent/browser/BoltAgent.ts:
```diff
@@ -0,0 +1,21 @@
+/**
```
the file names are created lowercase and separated by `-`. So, this file should be called `bolt-agent.ts` (or move this implementation to be directly in the `index.ts`)

bigmontz commented on packages/neo4j-driver-deno/generate.ts:
```diff
@@ -129,15 +129,15 @@ async function copyAndTransform(inDir: string, outDir: string) {
+      if (inPath.endsWith("channel/index.js") || inPath.endsWith("bolt-agent/index.ts")) {
         contents = contents.replace(
-          `export * from './node/index.js'`,
-          `export * from './deno/index.js'`,
+          `export * from './node/index'`,
+          `export * from './deno/index'`,
         );
```
Did you test the deno driver in testkit? I don't know if the `export` replace will work without the extension.

bigmontz commented on packages/neo4j-driver-lite/src/index.ts:
```diff
@@ -376,7 +379,7 @@ async function hasReachableServer (url: string, config?: Pick<Config, 'logging'>
 }
 
-const USER_AGENT: string = 'neo4j-javascript/' + VERSION
+const BOLT_AGENT = internal.boltAgent.BoltAgent.fromVersion('neo4j-javascript/' + VERSION)
```
Maybe move the constant definition before it usage. And I think this line might be `internal.boltAgent.fromVersion`

bigmontz commented on packages/neo4j-driver/src/index.js:
```diff
@@ -342,7 +343,7 @@ async function hasReachableServer (url, config) {
 }
 
-const USER_AGENT = 'neo4j-javascript/' + VERSION
+const BOLT_AGENT = internal.boltAgent.fromVersion('neo4j-javascript/' + VERSION)
```
Maybe move the definition to done before the usage (i know this was already this way, but I think it worth the small refactory).

bigmontz reviewed (commented):
This is looking good so far. 

I think it might worth added tests for old protocols to check if they are using the `boltAgent` as fallback for the case the user is not setting the `userAgent`.

ConorNeo committed:
Fix useragent string on metadata object in 5.3

ConorNeo committed:
changes made after rebase onto re-auth

ConorNeo committed:
fixed deno bolt agent

ConorNeo committed:
Browser node agent

ConorNeo committed:
browser bolt agent test

ConorNeo committed:
Add deno test

bigmontz committed:
Fix unit/integration tests related with connection.connect and bolt-agent for Deno

The test code related with `connection.connect` were calling the method without the bolt agent causing all kind of errors.
Add the bolt agent  to the calls and expectations solves the issue.

`Deno.osRelease()` and the flag `--allow-sys` are not available in Deno 1.19.2.
The missing flag was causing the test don't even start.
The method unexistance was causing the code fails because is calling a non existing method.

The missing flag is solved by using `--allow-all` instead.
This is a more permissive flag, but it is not a big deal in test env.

The absence of `Deno.osRelease()` is solved by not setting the release when this method is not available.
This might be an issue for old Deno users, but it's not a blocker.

ConorNeo committed:
Merge pull request #1 from bigmontz/bolt-agent

Fix unit/integration tests related with connection.connect and bolt-agent for Deno

bigmontz committed:
Revert changes in the `userAgent`

This property should not change the behaviour for backwards compatibility.
The `userAgent` appears on the logs and it might be parsed by reporters.

bigmontz committed:
Improve Deno instructions

ConorNeo committed:
Merge pull request #2 from bigmontz/bolt-agent

Revert changes in the `userAgent`

bigmontz committed:
Implements BoltAgent as structure

The BoltAgent was refactoried as structure to improve the log parser.

ConorNeo committed:
Merge pull request #3 from bigmontz/bolt-agent

Implements BoltAgent as structure

bigmontz committed:
Handling null/undefined platform information in Browser

bigmontz committed:
Use user agent instead of appVersion

ConorNeo committed:
Merge pull request #4 from bigmontz/bolt-agent

Handling null/undefined platform information in Browser

bigmontz committed:
Fix browser tests in the lite driver

The basic browser tests in the lite driver is done by using a WebSocket implementation in a mocked browser.
So, mocking the `window` is needed.

bigmontz committed:
Remove `boltAgent` from API and fix declaration tests

`boltAgent` is not configurable, so it should not be exposed in the `Config`.
The `InternalConfig` interface was created for keeping the type-safety.

The declaration tests were failing because of multiple `AbortSignal` definitons.
This was happening because each project is using different versions of `@types/node` in the `package-lock.json`.
The problem was not happening before because any interface exposed by the `neo4j-driver-core` was depending in Node defintions.
The solution can be done or by remove the Node api  reference or by pin the  `@types/node` versions accross all packages.
The first solution was choosed since the use of `string` instead `NodeJS.Platform` in the `SystemInfo` doesn't impact the usability of the `node.boltAgent.fromVersion` function.

ConorNeo committed:
Merge pull request #5 from bigmontz/bolt-agent

Fix browser tests in the lite driver

bigmontz committed:
Fix browser test in Node10 environment

The window mock was broken because it miss `Symbol`, `Date` and other global definitions.
Set the window as `globalThis` and then define the navigator in it solves the issue.

A proper way of solving the issue will be use `jsdom`, however this plugin doesn't work in Node10.

Also fixed an issue with timers getting hanging in some situations. The issue was catched by browser test with `--detectOpenHandles`.

ConorNeo committed:
Merge pull request #6 from bigmontz/bolt-agent

Fix browser test in Node10 environment

bigmontz reviewed (approved):
🥇
